src/twisted/web/_websocket_impl.py

```diff
@@ -0,0 +1,457 @@
+
+    def connectionLost(self, reason: Failure) -> None:
+        """
+        The websocket connection was lost.
+        """
+
+
+_WSP = TypeVar("_WSP", covariant=True, bound=WebSocketProtocol)
+_Bootstrap = Callable[[Union[WSConnection, Connection], ITransport], None]
+_log = Logger()
+
+
+class WebSocketServerFactory(TypingProtocol[_WSP]):
+    """
+    A L{WebSocketServerFactory} is a factory for a particular kind of
+    L{WebSocketProtocol} that implements server-side websocket listeners via
+    L{WebSocketResource}.
+    """
+
+    def buildProtocol(self, request: Request) -> _WSP:
+        """
+        To conform to L{WebSocketServerFactory}, you must implement a
+        C{buildProtocol} method which takes a L{Request
+        <twisted.web.server.Request>} and returns a L{WebSocketProtocol}.
+
+        @return: a L{WebSocketProtocol} that will handle the inbound
+            connection.
+        """
+
+
+class WebSocketClientFactory(TypingProtocol[_WSP]):
+    """
+    A L{WebSocketClientFactory} is a factory for a particular kind of
+    L{WebSocketProtocol} that implements client-side websocket listeners via
+    L{WebSocketClientEndpoint}.
+    """
+
+    def buildProtocol(self, url: str) -> _WSP:
+        """
+        To conform to L{WebSocketServerFactory}, you must implement a
+        C{buildProtocol} method which takes a string representing an URL and
+        returns a L{WebSocketProtocol}.
+
+        @return: a L{WebSocketProtocol} that will handle the outgoing
+            connection.
+        """
+
+
+@dataclass(frozen=True)
+class WebSocketClientEndpoint:
+    """
+    A L{WebSocketClientEndpoint} describes an URL to connect to and a way of
+    connecting to that URL, that can connect a L{WebSocketClientFactory} to
+    that URL.
+    """
+
+    endpointFactory: IAgentEndpointFactory
+    """
+    an L{IAgentEndpointFactory} that constructs agent endpoints when L{connect
+    <WebSocketClientEndpoint.connect>}
+    """
+    url: str
+    """
+    the URL to connect to.
+    """
+
+    @classmethod
+    def new(
+        cls,
+        reactor: IReactorTCP,
+        url: str,
+        tlsPolicy: IPolicyForHTTPS = BrowserLikePolicyForHTTPS(),
+        connectTimeout: int | None = None,
+        bindAddress: bytes | None = None,
+    ) -> WebSocketClientEndpoint:
+        """
+        Construct a L{WebSocketClientEndpoint} from a reactor and a URL.
+
+        @param reactor: The reactor to use for the TCP connection.
+
+        @param url: a string describing an URL where a websocket server lives.
+
+        @param tlsPolicy: The TLS policy to use for HTTPS connections.
+
+        @param connectTimeout: The number of seconds for the TCP-level
+            connection timeout.
+
+        @param bindAddress: The bind address to use for the TCP client
+            connections.
+
+        @return: the newly constructed endpoint.
+        """
+        sef = _StandardEndpointFactory(reactor, tlsPolicy, connectTimeout, bindAddress)
```

Comment:
To make the code easier to read, maybe use

```suggestion
        endpointFactory = _StandardEndpointFactory(reactor, tlsPolicy, connectTimeout, bindAddress)
```

Reply:
good catch